docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+spec:
+  trainingRuntimeRef:
+    name: pytorch-distributed-gpu
+    kind: ClusterTrainingRuntime
+  trainerConfig:
+    image: docker.io/custom-training
+  podSpecOverrides:
+    - targetReplicatedJobs:
+        - initializer
+          node
+      containers:
+        - name: user-identity
+          value: 123
+        - name: trainer
+          volumeMounts:
+            - name: user-123-volume
+              mountPath: /workspace
+      volumes:
+        - name: user-123-volume
+          persistentVolumeClaim:
+            claimName: user-123-volume
+```
+
+## The Training Runtime API
+
+The `TrainingRuntime` is the pre-created configurations of model training on the cluster,
+representing as blueprints. For example, Elastic PyTorch training, MPI DeepSpeed configuration,
+BERT LLM Fine-Tuning.
+
+These blueprints can be deployed within the Training Operator control plane and stored in a Kubeflow
+public repository that users can apply to their clusters.
+
+Platform or ML engineers can tweak existing blueprint, based on their requirements. For example,
+using custom configurations.
+
+The Kubeflow Training Operator can maintain more Training Runtimes when the community is ready to
+support them. For example, runtimes for [Jax](https://jax.readthedocs.io/en/latest/index.html) or
+[MLX](https://ml-explore.github.io/mlx/build/html/index.html). Initially, we will support: PyTorch,
+MPI, TensorFlow, XGBoost, and PaddlePaddle.
+
+The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`
+must be created and then changing the `TranJob` to point to the new version.
+This provides control as to how changes to runtimes propagate to existing training jobs.
+For example, when training is running for a long time (e.g. 1-2 months).
+
+In the future implementation, we will introduce a revision control mechanism similar to
+[Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#updating-a-deployment)
+to control versions of `TrainingRuntime` and enable rolling updates.
+
+We are going to create two CRDs: `TrainingRuntime` and `ClusterTrainingRuntime`. These runtimes have
+exactly the same APIs, but the first one is the namespace-scoped, the second is the cluster-scoped.
+If `trainingRuntimeRef` in `TrainJob` has the namespace, controller will use the `TrainingRuntime`,
+otherwise it will use the `ClusterTrainingRuntime`.
+
+```golang
+type TrainingRuntime struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Framework specific parameters.
+    MLSpec *MLSpec `json:"mlSpec,omitempty"`
+
+    // Number of nodes to execute training.
+    NumNodes int `json:"numNodes,omitempty"`
+
+    // JobSet spec.
+    JobSetSpec *batchv1.JobSetSpec `json:",inline"`
+
+    // For gang-scheduling using volcano or scheduler plugins, supported for all frameworks.
+    GangScheduler *kubeflowv1.GangScheduler `json:"gangScheduler,omitempty"`
```

Comment:
Just a curiosity, for the ML user, are they more familiar with Gang scheduling? I know gang is common in the developer community, but how far does that extend? Why not call it GroupScheduling?

Reply:
Usually, MLOps engineers should be aware of these terms, but I let @tenzen-y to speak more on this.

Reply:
@tenzen-y @vsoch Should we just call it `scheduler` to make it more generic?

Reply:
I think that `gangScheduler` should be fine since this is not related to scheduler field to PodSpec.
This is only for PodGroup reconciliations. The `scheduler` field name could bring users some confusions.

Reply:
I am wondering, do we want to re-use this parameter not only for gang-scheduling ?

For example, if users want to use custom scheduler (YuniKorn or Volcano) to schedule TrainJobs with 1 node ?
For example, Spark Operator introduces YuniKorn support recently: https://github.com/kubeflow/spark-operator/pull/2107 (cc @yuchaoran2011 @vara-bonthu @shravan-achar @jacobsalway @ChenYi015)

Could it be useful for ML Training Jobs as well ?

Reply:
We can consider using this API to change pod scheduler: https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3874